pkg/morph/deploy/contracts.go

```diff
@@ -1,7 +1,410 @@
+	}
+
+	localVersion, err := readContractLocalVersion(prm.blockchain, prm.committee, prm.localNEF, prm.localManifest, prm.anyValidExtraDeployArgs...)
+	if err != nil {
+		return util.Uint160{}, fmt.Errorf("read version of the local contract: %w", err)
+	}
+
+	// wrap the parent context into the context of the current function so that
+	// transaction wait routines do not leak
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	setGroupInManifest(&prm.localManifest, prm.localNEF, prm.committeeGroupKey, prm.localAcc.ScriptHash())
+
+	var managementContract *management.Contract
+	if prm.committeeDeployRequired {
+		deployCommitteeActor, err := newCommitteeNotaryActorWithScope(prm.blockchain, prm.localAcc, prm.committee, transaction.Global)
+		if err != nil {
+			return util.Uint160{}, fmt.Errorf("create Notary service client sending deploy transactions to be signed by the committee: %w", err)
+		}
+
+		managementContract = management.New(deployCommitteeActor)
+	} else {
+		managementContract = management.New(localActor)
+	}
+
+	var alreadyUpdated bool
+	domainNameForAddress := prm.domainName + "." + domainContractAddresses
+	l := prm.logger.With(zap.String("contract", prm.localManifest.Name), zap.String("domain", domainNameForAddress))
+	deployTxMonitor := newTransactionGroupMonitor(localActor)
+	updateTxMonitor := newTransactionGroupMonitor(localActor)
+	registerDomainTxMonitor := newTransactionGroupMonitor(localActor)
+	registerTLDTxMonitor := newTransactionGroupMonitor(localActor)
+	setDomainRecordTxMonitor := newTransactionGroupMonitor(localActor)
+
+	for ; ; prm.monitor.waitForNextBlock(ctx) {
+		select {
+		case <-ctx.Done():
+			return util.Uint160{}, fmt.Errorf("wait for the contract synchronization: %w", ctx.Err())
+		default:
+		}
+
+		l.Info("reading on-chain state of the contract by NNS domain name...")
+
+		var missingDomainName, missingDomainRecord bool
+
+		onChainState, err := readContractOnChainStateByDomainName(prm.blockchain, prm.nnsContract, domainNameForAddress)
+		if err != nil {
+			if errors.Is(err, errContractNotFound) {
+				l.Error("contract is recorded in the NNS but not found on the chain, will wait for a background fix")
+				continue
+			}
+
+			missingDomainName = errors.Is(err, errMissingDomain)
+			if !missingDomainName {
+				missingDomainRecord = errors.Is(err, errMissingDomainRecord)
+				if !missingDomainRecord {
+					if errors.Is(err, errInvalidContractDomainRecord) {
+						l.Error("contract's domain record is invalid/unsupported, will wait for a background fix", zap.Error(err))
+					} else {
+						l.Error("failed to read on-chain state of the contract record by NNS domain name, will try again later", zap.Error(err))
+					}
+					continue
+				}
+			}
+
+			l.Error("could not read on-chain state of the contract by NNS domain name, trying by pre-calculated address...")
+
+			preCalculatedAddr := state.CreateContractHash(prm.localAcc.ScriptHash(), prm.localNEF.Checksum, prm.localManifest.Name)
```

Comment:
i think yes: transaction sender should always be the same acc (to keep hash predictable), so, even if committee multisig is required, we calculate hash from simple acc. What do u think? 

btw `Actor` provides `Sender()` interface, so it may be useful here

Reply:
done